lib/vrl/stdlib/src/tag_types_externally.rs

```diff
@@ -0,0 +1,225 @@
+use shared::btreemap;
+use std::collections::BTreeMap;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct TagTypesExternally;
+
+impl Function for TagTypesExternally {
+    fn identifier(&self) -> &'static str {
+        "tag_types_externally"
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[
+            Example {
+                title: "scalar",
+                source: "tag_types_externally(123)",
+                result: Ok(r#"{ "integer": 123 }"#),
+            },
+            Example {
+                title: "object",
+                source: indoc! {r#"
+                    tag_types_externally({
+                        "message": "Hello world",
+                        "request": {
+                            "duration_ms": 67.9
+                        }
+                    })
+                "#},
+                result: Ok(
+                    r#"{ "message": { "bytes": "Hello world" }, "request": { "duration_ms": { "float": 67.9 } } }"#,
+                ),
+            },
+            Example {
+                title: "array",
+                source: r#"tag_types_externally(["foo", "bar"])"#,
+                result: Ok(r#"[{ "bytes": "foo" }, { "bytes": "bar" }]"#),
+            },
+        ]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+
+        Ok(Box::new(TagTypesExternallyFn { value }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::ANY,
+            required: true,
+        }]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct TagTypesExternallyFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for TagTypesExternallyFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let value = self.value.resolve(ctx)?;
+        let tagged_externally = tag_type_externally(value);
+
+        Ok(tagged_externally)
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        match self.value.type_def(state).kind() {
+            kind if kind.is_array() => TypeDef::new()
+                .infallible()
+                .array_mapped::<(), Kind>(map! { (): Kind::all() }),
+            _ => TypeDef::new()
+                .infallible()
+                .object::<(), Kind>(map! { (): Kind::all() }),
+        }
+    }
+}
+
+fn tag_type_externally(value: Value) -> Value {
+    let (key, value) = match value {
+        value @ Value::Bytes(_) => (Some("bytes"), value),
```

Comment:
Externally, `bytes` is known as `string`, so we should tag with that instead.

Reply:
Oh, I see this was discussed, but not resolved. I still cast my vote towards `string`, you are correct in that it is not guaranteed to be valid utf8, but for all practical purposes I think we can assume it is for this..

Reply:
The issue description has an example with a `"string"` label, so I don't think it implied not to tag it.

Will change it to `"string"` now.